File: go/vehicle-purchase/vehicle_purchase.go
package purchase

// NeedsLicense determines whether a license is needed to drive a type of vehicle. Only "car" and "truck" require a license.
func NeedsLicense(kind string) bool {
	if kind == "car" || kind == "truck" {
		return true
	}
	return false
}

// ChooseVehicle recommends a vehicle for selection. It always recommends the vehicle that comes first in lexicographical order.
func ChooseVehicle(option1, option2 string) string {
	var bestOption string

	if option1 < option2 {
		bestOption = option1
	} else {
		bestOption = option2
	}

	return bestOption + " is clearly the better choice."
}

// CalculateResellPrice calculates how much a vehicle can resell for at a certain age.
func CalculateResellPrice(originalPrice, age float64) float64 {
	var percentageReduction float64

	if age < 3.0 {
		percentageReduction = 0.80
	} else if age >= 10.0 {
		percentageReduction = 0.50
	} else {
		percentageReduction = 0.70
	}

	return originalPrice * percentageReduction
}

File: javascript/vehicle-purchase/vehicle-purchase.js
// @ts-check
//
// The line above enables type checking for this file. Various IDEs interpret
// the @ts-check directive. It will give you helpful autocompletion when
// implementing this exercise.

/**
 * Determines whether or not you need a licence to operate a certain kind of vehicle.
 *
 * @param {string} kind
 * @returns {boolean} whether a license is required
 */
export function needsLicense(kind) {
  const licensedVehicles = ["car", "truck"]
  return licensedVehicles.includes(kind)
}

/**
 * Helps choosing between two options by recommending the one that
 * comes first in dictionary order.
 *
 * @param {string} option1
 * @param {string} option2
 * @returns {string} a sentence of advice which option to choose
 */
export function chooseVehicle(option1, option2) {
  let preference = (option1 < option2) ? option1 : option2

  return preference + " is clearly the better choice."
}

/**
 * Calculates an estimate for the price of a used vehicle in the dealership
 * based on the original price and the age of the vehicle.
 *
 * @param {number} originalPrice
 * @param {number} age
 * @returns expected resell price in the dealership
 */
export function calculateResellPrice(originalPrice, age) {
  let reductionPercentage = 0.0
  
  if (age < 3) {
    reductionPercentage = 0.20
  } else if (age > 10) {
    reductionPercentage = 0.50
  } else {
    reductionPercentage = 0.30
  }

  const reductionAmount = originalPrice * reductionPercentage
  return originalPrice - reductionAmount
}
